### PR TITLE
Fenda+Ditto model

### PR DESCRIPTION
**Pull request by Justinyangjy**
# PR Type
[**Feature** | Fix | Documentation | Other ]

# Short Description

[Clickup Ticket(s): Link.](https://app.clickup.com/t/86880yyex)

Added Fenda+Ditto model

# Tests Added

added smoke tests and Fenda+Ditto model tests in terms of parameter swapping and loss calculation

### CODE REVIEW

File: tests/clients/test_fenda_ditto_client.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    for layer_init_global_tensor, layer_params in zip(fenda_ditto_client.initial_global_tensors, params):
+        assert pytest.approx(torch.sum(layer_init_global_tensor - layer_params), abs=0.0001) == 0.0
+    # Make sure the global model was set correctly
+    for global_model_layer_params, layer_params in zip(global_model_params, params):
+        assert pytest.approx(torch.sum(global_model_layer_params - layer_params), abs=0.0001) == 0.0
+    # # Make sure the local model was set correctly
+    for local_model_layer_params, layer_params in zip(fenda_model_params, params):
+        assert pytest.approx(torch.sum(local_model_layer_params - layer_params), abs=0.0001) == 0.0
+
+    # # Now we update in a later round
```
Comment (emersodb) [inline tests/clients/test_fenda_ditto_client.py:51]: Double comment guy 🙂

File: tests/clients/test_fenda_ditto_client.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    config: Config = {"current_server_round": 1}
+
+    params = [val.cpu().numpy() + 1.0 for _, val in fenda_ditto_client.global_model.state_dict().items()]
+    fenda_ditto_client.set_parameters(params, config, fitting_round=True)
+    fenda_ditto_client.update_before_train(1)
+
+    global_model_params = [
+        val.detach().clone()
+        for name, val in fenda_ditto_client.global_model.state_dict().items()
+        if name.startswith("base_module.")
```
Comment (emersodb) [inline tests/clients/test_fenda_ditto_client.py:28]: With the initialize_all_weights functionality, we should match the params for all components of the `global_model_params` not just those in the base_module, if I'm not mistaken?

File: tests/clients/test_fenda_ditto_client.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+    # # Now we update in a later round
+    config = {"current_server_round": 2}
+    params = [param + 1.0 for param in params]
+    fenda_ditto_client.set_parameters(params, config, fitting_round=True)
+    fenda_ditto_client.update_before_train(2)
+    global_model_params = [
+        val.detach().clone()
+        for name, val in fenda_ditto_client.global_model.state_dict().items()
+        if name.startswith("base_module.")
```
Comment (emersodb) [inline tests/clients/test_fenda_ditto_client.py:59]: With a full parameter exchanger, we should match the params for all components of the `global_model_params` not just those in the base_module, if I'm not mistaken?

File: tests/clients/test_fenda_ditto_client.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    fenda_ditto_client.model.first_feature_extractor.load_state_dict(state_dict)
+    fenda_ditto_client.model.second_feature_extractor.load_state_dict(state_dict)
+
+    preds = {"global": torch.tensor([[1.0, 0.0], [0.0, 1.0]]), "local": torch.tensor([[1.0, 0.0], [0.0, 1.0]])}
+    target = torch.tensor([[1.0, 0.0], [1.0, 0.0]])
+    training_loss = fenda_ditto_client.compute_training_loss(preds, {}, target)
+    fenda_ditto_client.global_model.eval()
+    fenda_ditto_client.model.eval()
+    evaluation_loss = fenda_ditto_client.compute_evaluation_loss(preds, {}, target)
+    print(evaluation_loss)
```
Comment (emersodb) [inline tests/clients/test_fenda_ditto_client.py:133]: Dropping print statements in tests is good practice 🙂

File: tests/clients/test_fenda_ditto_client.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        state_dict[key] += 0.1
+    fenda_ditto_client.model.first_feature_extractor.load_state_dict(state_dict)
+
+    preds = {"global": torch.tensor([[1.0, 0.0], [0.0, 1.0]]), "local": torch.tensor([[1.0, 0.0], [0.0, 1.0]])}
+    target = torch.tensor([[1.0, 0.0], [1.0, 0.0]])
+    training_loss = fenda_ditto_client.compute_training_loss(preds, {}, target)
+    fenda_ditto_client.global_model.eval()
+    fenda_ditto_client.model.eval()
+    evaluation_loss = fenda_ditto_client.compute_evaluation_loss(preds, {}, target)
+    print(evaluation_loss)
```
Comment (emersodb) [inline tests/clients/test_fenda_ditto_client.py:167]: Dropping print statements in tests is good practice 🙂

File: tests/clients/test_fenda_ditto_client.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+import torch
+from flwr.common import Config
+
+from fl4health.clients.fenda_ditto_client import FendaDittoClient
+from fl4health.model_bases.fenda_base import FendaModel
+from fl4health.model_bases.sequential_split_models import SequentiallySplitExchangeBaseModel
+from fl4health.parameter_exchange.full_exchanger import FullParameterExchanger
+from tests.clients.fixtures import get_client  # noqa
+from tests.test_utils.models_for_test import FeatureCnn, FendaHeadCnn, HeadCnn
+
```
Comment (emersodb) [inline tests/clients/test_fenda_ditto_client.py:11]: One final test that I would recommend is that we test that assertion of the model structures being the same. You can find an example of testing that an assertion raises in `test_get_parameter_exchanger_with_incorrect_model` if you've never done it before.

**Review by emersodb — COMMENTED** [review]
Overall, the flow of the PR looks good. There are a few refactors that should be beneficial and perhaps a test to add.

File: examples/fenda_ditto_example/client.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            freeze_global_feature_extractor=freeze_global_feature_extractor,
+        )
+
+        pre_aggregation_checkpointer = BestLossTorchCheckpointer(
+            checkpoint_dir, f"fenda_ditto_client_{self.client_name}_pre_agg.pkl"
+        )
+        post_aggregation_checkpointer = BestLossTorchCheckpointer(
+            checkpoint_dir, f"fenda_ditto_client_{self.client_name}_post_agg.pkl"
+        )
+        self.checkpointer = ClientCheckpointModule(
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply examples/fenda_ditto_example/client.py:60]: makes sense, I will delete this init and moved the checkpointer to `main`, I also added the argument for users to define which checkpointer to use, or both

File: fl4health/clients/fenda_ditto_client.py [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+        checkpointer: Optional[ClientCheckpointModule] = None,
+        lam: float = 1.0,
+        freeze_global_feature_extractor: bool = False,
+    ) -> None:
+        """
+        This client implements the Ditto algorithm from Ditto: Fair and Robust Federated Learning Through
+        Personalization with a local FENDA model. The idea is that we want to train a local FENDA model
+        along with the global model for each client. So we simultaneously train a global model that is
+        aggregated on the server-side and use those weights to also constrain the training of a local
+        FENDA model. The constraint for this local FENDA model is identical to the FedProx loss.
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py:36]: Good idea! This definitely makes things more clear

File: fl4health/clients/fenda_ditto_client.py [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+                each batch. Defaults to LossMeterType.AVERAGE.
+            checkpointer (Optional[ClientCheckpointModule], optional): Checkpointer module defining when and how to
+                do checkpointing during client-side training. No checkpointing is done if not provided. Defaults to
+                None.
+            metrics_reporter (Optional[MetricsReporter], optional): A metrics reporter instance to record the metrics
+                during the execution. Defaults to an instance of MetricsReporter with default init parameters.
+            lam (float, optional): weight applied to the Ditto drift loss. Defaults to 1.0.
+            freeze_global_feature_extractor (bool, optional): Determines whether we freeze the FENDA global feature
+                extractor during training. FendaDitto default is both the global and the local feature extractor in
+                the local FENDA model will be trained, the Ditto loss will be calculated using the local FENDA
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py:53]: sounds good!

File: fl4health/clients/fenda_ditto_client.py [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+        Set dataloaders, optimizers, parameter exchangers and other attributes derived from these.
+        Then set initialized attribute to True.
+
+        Args:
+            config (Config): The config from the server.
+        """
+        self.global_model = self.get_global_model(config).to(self.device)
+        super().setup_client(config)
+
+        # Check if shapes of self.global_model.feature_extractor and self.model.second_feature_extractor match
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py:133]: sounds good the function is `check_shape_match` in `fl4health/utils`in the file `parameter_extraction.py`

File: fl4health/clients/fenda_ditto_client.py [hunk 11]
```diff
@@ -0,0 +1,444 @@
...
+from fl4health.clients.basic_client import BasicClient, TorchInputType
+from fl4health.losses.weight_drift_loss import WeightDriftLoss
+from fl4health.model_bases.fenda_base import FendaModel
+from fl4health.model_bases.sequential_split_models import SequentiallySplitExchangeBaseModel
+from fl4health.parameter_exchange.full_exchanger import FullParameterExchanger
+from fl4health.utils.losses import EvaluationLosses, LossMeterType, TrainingLosses
+from fl4health.utils.metrics import Metric
+
+
+class FendaDittoClient(BasicClient):
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py:20]: This makes sense, I will look at where I can reduce my code with this change

File: fl4health/clients/fenda_ditto_client.py [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+
+        Returns:
+            NDArrays: GLOBAL model weights to be sent to the server for aggregation
+        """
+        assert self.global_model is not None and self.parameter_exchanger is not None
+        return self.parameter_exchanger.push_parameters(self.global_model, config=config)
+
+    def set_parameters(self, parameters: NDArrays, config: Config, fitting_round: bool) -> None:
+        """
+        The parameters being passed are to be routed to the global model copied to the global feature extractor
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py:165]: makes more sense 😆

File: fl4health/clients/fenda_ditto_client.py [hunk 13]
```diff
@@ -0,0 +1,444 @@
...
+
+    def initialize_all_model_weights(self, parameters: NDArrays, config: Config) -> None:
+        """
+        If this is the first time we're initializing the model weights, we initialize the global weights.
+
+        Args:
+            parameters (NDArrays): Model parameters to be injected into the client model
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
+        """
+        self.parameter_exchanger.pull_parameters(parameters, self.global_model, config)
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py:212]: I have modified the comments to explain what are are/aren't doing more clearly!

File: fl4health/clients/fenda_ditto_client.py [hunk 14]
```diff
@@ -0,0 +1,444 @@
...
+        if current_server_round == 1 and fitting_round:
+            log(INFO, "Initializing the global and local models weights for the first time")
+            self.initialize_all_model_weights(parameters, config)
+        else:
+            log(INFO, "Setting the global model weights")
+            self.parameter_exchanger.pull_parameters(parameters, self.global_model, config)
+        # GLOBAL MODEL feature extractor is given to local FENDA model
+        self.model.second_feature_extractor.load_state_dict(self.global_model.base_module.state_dict())
+
+    def update_before_train(self, current_server_round: int) -> None:
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py:194]: Done!

File: fl4health/clients/fenda_ditto_client.py [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+        """
+        If this is the first time we're initializing the model weights, we initialize the global weights.
+
+        Args:
+            parameters (NDArrays): Model parameters to be injected into the client model
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
+        """
+        self.parameter_exchanger.pull_parameters(parameters, self.global_model, config)
+
+    def train_by_epochs(
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py:214]: put the freezing all in `update_before_train`!

File: fl4health/clients/fenda_ditto_client.py [hunk 16]
```diff
@@ -0,0 +1,444 @@
...
+    def compute_training_loss(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> TrainingLosses:
+        """
+        Computes training losses given predictions of the global and local models and ground truth data.
+        For the local model we add to the vanilla loss function by including Ditto penalty loss which is the l2 inner
+        product between the initial global model feature extractor weights and feature extractor weights of the
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py:376]: Yes, I have changed this!

File: tests/clients/test_fenda_ditto_client.py (hunk 4 above)
Comment (Justinyangjy) [reply tests/clients/test_fenda_ditto_client.py:133]: removed!

File: tests/clients/test_fenda_ditto_client.py (hunk 5 above)
Comment (Justinyangjy) [reply tests/clients/test_fenda_ditto_client.py:167]: removed

File: tests/clients/test_fenda_ditto_client.py (hunk 6 above)
Comment (Justinyangjy) [reply tests/clients/test_fenda_ditto_client.py:11]: Added the test + corrected the assertion

File: tests/clients/test_fenda_ditto_client.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+    fenda_model_params = [
+        val.detach().clone()
+        for name, val in fenda_ditto_client.model.state_dict().items()
+        if name.startswith("second_feature_extractor.")
+    ]
+
+    # First fitting round we should set both the global and local models to params and store the global model values
+    assert fenda_ditto_client.initial_global_tensors is not None
+    # Tensors should be conv1 weights, biases, conv2 weights, biases, fc1 weights, biases (so 6 total)
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply tests/clients/test_fenda_ditto_client.py:39]: yes, changed to 4!

File: tests/clients/test_fenda_ditto_client.py (hunk 2 above)
Comment (Justinyangjy) [reply tests/clients/test_fenda_ditto_client.py:28]: yes! removing the `if...`

File: tests/clients/test_fenda_ditto_client.py (hunk 3 above)
Comment (Justinyangjy) [reply tests/clients/test_fenda_ditto_client.py:59]: yes! removing the `if...`

**Comment by Justinyangjy** [comment]
The changes should be good for review, however, the new added `test_setup_client_with_incorrect_model` isnt really elegant as there would be too many mocks if I did the actual check. doing the check before the setup will throw an error as the global model is not defined.
currently the setup_client itself is mocked with the check code. I would appreciate some pointers to make this better. @emersodb

File: fl4health/clients/fenda_ditto_client.py [hunk 18]
```diff
@@ -178,122 +160,42 @@ def set_parameters(self, parameters: NDArrays, config: Config, fitting_round: bo
...
-        assert self.parameter_exchanger is not None and isinstance(self.parameter_exchanger, FullParameterExchanger)
-
-        current_server_round = self.narrow_config_type(config, "current_server_round", int)
-        if current_server_round == 1 and fitting_round:
-            log(INFO, "Initializing the global and local models weights for the first time")
-            self.initialize_all_model_weights(parameters, config)
-        else:
-            log(INFO, "Setting the global model weights")
-            self.parameter_exchanger.pull_parameters(parameters, self.global_model, config)
+        super().set_parameters(parameters, config, fitting_round)
```
[the remark this answers is not in the document]
Comment (emersodb) [reply fl4health/clients/fenda_ditto_client.py:163]: Correct me if I'm wrong in that interpretation though.

**Review by emersodb — COMMENTED** [review]
Changes look good overall. Just a few additional comments out of the refactor and we should be good to go thereafter.

File: fl4health/clients/fenda_ditto_client.py [hunk 19]
```diff
@@ -372,28 +244,31 @@ def compute_training_loss(
...
         Computes training losses given predictions of the global and local models and ground truth data.
-        For the local model we add to the vanilla loss function by including Ditto penalty loss which is the l2 inner
-        product between the initial global model feature extractor weights and feature extractor weights of the
-        local model. This is stored in "backward". The loss to optimize the global model is stored in the
-        additional losses dictionary under "global_loss".
+        For the local model, we add to the vanilla loss function by including a Ditto penalty loss. This penalty
+        is the L2 inner product between the initial global model feature extractor weights and the feature extractor
+        weights of the local model. If the global feature extractor is not frozen, the penalty is computed using the
+        global feature extractor of the local model. If it is frozen, the penalty is computed using the local feature
+        extractor of the local model. allowing for flexibility in training scenarios where the feature extractors
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py]: Good catch, I changed the incomplete sentence to `This allows ...`

File: fl4health/clients/fenda_ditto_client.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py]: I think this change makes a lot of sense, I also did copy 
`
# Make sure that the proper components exist.
        assert self.global_model is not None and self.model is not None
        assert self.parameter_exchanger is not None and isinstance(self.parameter_exchanger, FullParameterExchanger)
 `
from the `ditto_client.py`  `set_parameters()`.

File: fl4health/clients/fenda_ditto_client.py [hunk 20]
```diff
@@ -178,122 +160,42 @@ def set_parameters(self, parameters: NDArrays, config: Config, fitting_round: bo
...
-        Args:
-            epochs (int): The number of epochs for local training.
-            current_round (Optional[int]): The current FL round.
-
-        Returns:
-            Tuple[Dict[str, float], Dict[str, Scalar]]: The loss and metrics dictionary from the local training.
-                Loss is a dictionary of one or more losses that represent the different components of the loss.
-        """
+    def update_before_train(self, current_server_round: int) -> None:
+        # freeze the global freature extractor during training updates if true.
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py]: haha yes!

File: fl4health/clients/fenda_ditto_client.py [hunk 21]
```diff
@@ -119,6 +98,21 @@ def get_global_model(self, config: Config) -> SequentiallySplitExchangeBaseModel
         """
         raise NotImplementedError
 
+    def _check_shape_match(self) -> None:
+        # Check if shapes of self.global_model.feature_extractor and self.model.second_feature_extractor match
+        check_shape_match(
+            self.global_model.base_module.parameters(),
+            self.model.second_feature_extractor.parameters(),
+            "Shapes of self.global_model.feature_extractor and self.model.second_feature_extractor do not match.",
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py]: added!

File: fl4health/clients/fenda_ditto_client.py [hunk 22]
```diff
@@ -33,7 +32,9 @@ def __init__(
         Personalization with a local FENDA model. The idea is that we want to train a local FENDA model
         along with the global model for each client. So we simultaneously train a global model that is
         aggregated on the server-side and use those weights to also constrain the training of a local
-        FENDA model. The constraint for this local FENDA model is identical to the FedProx loss.
+        FENDA model. The feature extractor from globally aggregated model is, at each server round,
+        injected into the global feature extractor of the FENDA model. The constraint for this local
+        FENDA model uses a weight drift loss on its feature extraction models.
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/fenda_ditto_client.py]: got it!

File: fl4health/clients/ditto_client.py [hunk 23]
```diff
@@ -89,7 +92,7 @@ def setup_client(self, config: Config) -> None:
         """
         # Need to setup the global model here as well. It should be the same architecture as the local model so
```
[the remark this answers is not in the document]
Comment (Justinyangjy) [reply fl4health/clients/ditto_client.py:95]: done!

**Review by emersodb — APPROVED** [review]
Looks good to me!

**Pull request merged**